api/agent/agent.go

```diff
@@ -583,24 +592,54 @@ func (s *hotSlot) exec(ctx context.Context, call *call) error {
 	defer stdoutWrite.Close()
 
 	proto := protocol.New(protocol.Protocol(call.Format), stdinWrite, stdoutRead)
-
 	swapBack := s.container.swap(stdinRead, stdoutWrite, call.stderr, &call.Stats)
 	defer swapBack() // NOTE: it's important this runs before the pipes are closed.
 
 	errApp := make(chan error, 1)
 	go func() {
 		ci := protocol.NewCallInfo(call.Call, call.req)
-		errApp <- proto.Dispatch(ctx, ci, call.w)
+		err := proto.Dispatch(ctx, ci, call.w)
+
+		// drain the container stdout if bad gw (aka container communication failed)
+		if err != nil && models.GetAPIErrorCode(err) == http.StatusBadGateway {
+			go func() {
+				// stdoutWrite.Close() will unblock this with EOF incoming from stdoutRead
+				io.Copy(ioutil.Discard, stdoutRead)
```

Comment:
idear: can we do this in the reader when we exceed the max instead of the contract requiring the caller to do this (see this at least 2 places)